internal/backend/runtime/omni/controllers/omni/infra_machine.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/infra"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/siderolink"
+)
+
+// InfraMachineController manages InfraMachine resource lifecycle.
+//
+// InfraMachineController transforms an Omni Machine managed by a static infra provider to an infra.Machine, applying the user overrides in omni.InfraMachineConfig resource if present.
+type InfraMachineController = qtransform.QController[*siderolink.Link, *infra.Machine]
+
+// NewInfraMachineController initializes InfraMachineController.
+func NewInfraMachineController() *InfraMachineController {
+	return qtransform.NewQController(
+		qtransform.Settings[*siderolink.Link, *infra.Machine]{
+			Name: "InfraMachineController",
+			MapMetadataFunc: func(link *siderolink.Link) *infra.Machine {
+				return infra.NewMachine(link.Metadata().ID())
+			},
+			UnmapMetadataFunc: func(infraMachine *infra.Machine) *siderolink.Link {
+				return siderolink.NewLink(resources.DefaultNamespace, infraMachine.Metadata().ID(), nil)
+			},
+			TransformFunc: func(ctx context.Context, r controller.Reader, _ *zap.Logger, link *siderolink.Link, infraMachine *infra.Machine) error {
+				infraMachine.TypedSpec().Value.Connected = link.TypedSpec().Value.Connected
+
+				providerID, ok := link.Metadata().Annotations().Get(omni.LabelInfraProviderID)
+				if !ok {
+					return xerrors.NewTaggedf[qtransform.SkipReconcileTag]("the link is not created by an infra provider")
+				}
+
+				providerStatus, err := safe.ReaderGetByID[*infra.ProviderStatus](ctx, r, providerID)
+				if err != nil {
+					return err
+				}
+
+				if _, isStaticProvider := providerStatus.Metadata().Labels().Get(omni.LabelIsStaticInfraProvider); !isStaticProvider {
+					return xerrors.NewTaggedf[qtransform.SkipReconcileTag]("the link is not created by a static infra provider")
+				}
+
+				infraMachine.Metadata().Labels().Set(omni.LabelInfraProviderID, providerID)
+
+				clusterMachineConfigStatus, err := safe.ReaderGetByID[*omni.ClusterMachineConfigStatus](ctx, r, link.Metadata().ID())
+				if err != nil && !state.IsNotFoundError(err) {
+					return err
+				}
+
+				if clusterMachineConfigStatus != nil {
+					infraMachine.TypedSpec().Value.TalosVersion = clusterMachineConfigStatus.TypedSpec().Value.TalosVersion
+					infraMachine.TypedSpec().Value.Allocated = true
+
+					if clusterMachineConfigStatus.TypedSpec().Value.LastConfigError == "" { // we observed that a config was successfully applied at least once - assume installed
+						infraMachine.TypedSpec().Value.Installed = true // todo: probably not reliable
+					}
+				} else {
+					infraMachine.TypedSpec().Value.TalosVersion = ""
+					infraMachine.TypedSpec().Value.Allocated = false
+					infraMachine.TypedSpec().Value.Installed = false // todo: probably not reliable, how do we actually know if it's installed or not?
```

Comment:
To the reviewer: I need some good ideas here. I basically want to reliably detect if Talos is installed on the disk or not. This affects the wipe decision on the provider.